Xiaomi vacuum as platform of new `vacuum` component derived from ToggleEntity, and services

azogue opened the pull request:
## Description:

Conversion for the **Xiaomi Vacuum cleaner robot** from a platform switch to a **`vacuum` platform** (NEW COMPONENT for botvacs), as a toggle entity, sensors and services.

Proposal to the conversation maintained [here](https://github.com/home-assistant/home-assistant/pull/8416#issuecomment-317087533), to better integrate in Home Assistant the fantastic work done by @rytilahti to control this awesome cleaning robot.

(**UPDATED**) Changes:
- Conversion from switch platform to vacuum platform (**BREAKING-CHANGE**)
- Add services proposed in #8416 by @bramkragten to the new component, with shorter names.
- Add attributes for the vacuum robot to be handled in the custom UI.
- Assign default icons to sensors and to the switch (`mdi:google-circles-group` looks like the robot!)
- Common services are: `turn_on`, `turn_off`, `start_pause`, `stop`, `return_to_base`, `locate`, `set_fanspeed` and `send_command`.
- The other services (for remote control the robot) are implemented as platform dependent, so they start with `xiaomi_`: `'xiaomi_remote_control_move`, `xiaomi_remote_control_move_step`, `xiaomi_remote_control_start` and `xiaomi_remote_control_stop`.
- ~The associated sensors are also optional, so it's an each platform-decision, along with the map-cameras (in the neato vacuum)~
- `SUPPORT_*` flags for each service to render a custom UI. This way, there is no need to implement all common services for each botvac.

**Pull request in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io) with documentation:** home-assistant/home-assistant.github.io#3076

**Pull request in [home-assistant-polymer](https://github.com/home-assistant/home-assistant-polymer):** home-assistant/home-assistant-polymer#359

## Example entry for `configuration.yaml` (if applicable):
```yaml
vacuum:
- platform: xiaomi
  name: Vacuum cleaner
  host: !secret xiaomi_vacuum_robot_ip
  token: !secret xiaomi_vacuum_robot_token
```

## Checklist:

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io)

If the code communicates with devices, web services, or third-party tools:
  - [x] Local tests with `tox` run successfully. **Your PR cannot be merged unless tests pass**
  - [x] New dependencies are only imported inside functions that use them ([example][ex-import]).
  - [x] New dependencies have been added to `requirements_all.txt` by running `script/gen_requirements_all.py`.
  - [x] New files were added to `.coveragerc`.

[ex-import]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L54

azogue committed:
Xiaomi vacuum as component with switch, sensors and services

- Conversion from switch platform to async component.
- Add services proposed in #8416 to the new component, with shorter names.
- Add sensors for the vacuum robot as a selectable list from `battery`, `state`, `error`, `fanspeed`, `clean_time` and `clean_area` (the state attributes of the switch). The sensors don't poll, but listen to a signal to update the state, the switch fires this signal when updating.
- Assign default icons to sensors and the switch (`mdi:google-circles-group` looks like the robot!)

azogue committed:
path change in requirements_all (from switch platform to component)

azogue committed:
copy pasting is a bad habit

azogue committed:
services to the components services.yaml, modify .coveragerc

pvizeli commented on homeassistant/components/sensor/xiaomi_vacuum.py:
```diff
@@ -0,0 +1,28 @@
+    sensors = discovery_info[CONF_SENSORS]
+    mirobo_vacuum = hass.data[DOMAIN]
+
+    yield from async_add_devices(
```
that is only a function no coroutine

azogue replied:
oops, sorry! I'll fix it

pvizeli commented on homeassistant/components/switch/xiaomi_vacuum.py:
```diff
@@ -4,121 +4,23 @@
-            self._is_on = state.is_on
-        except DeviceException as ex:
-            _LOGGER.error("Got exception while fetching the state: %s", ex)
+    yield from async_add_devices(
```
Not a coroutine

azogue replied:
same here, thanks

pvizeli reviewed (changes requested):
If it possible to handle remotes with as platform of remotes?

azogue committed:
review: use with multiple hosts, fix calls to async_add_devices, fix ranges for services

azogue commented:
@pvizeli, do you mean handling multiple vacuum cleaners, right? I'm changing it to allow handling more than one robot. The yaml config for 1 robot is the same, for multiple hosts should look like this:

```yaml
xiaomi_vacuum:
- name: Vacuum cleaner first floor
  host: !secret xiaomi_vacuum_robot_ip_1
  token: !secret xiaomi_vacuum_robot_token_1
  sensors:
    - state
    - battery
    - fanspeed
    - clean_area
    - clean_time
    - error

- name: Vacuum cleaner second floor
  host: !secret xiaomi_vacuum_robot_ip_2
  token: !secret xiaomi_vacuum_robot_token_2
  sensors:
    - state
    - battery
    - fanspeed
    - clean_area
```

Also, I added `host` to the services schemas, being optional and defaulting to use all vacuum cleaners. 
I did not use the `entity_id` parameter for picking out the robots because that unique entity for these devices is a `switch`, but I could change it, what do you think about?
(for each vacuum cleaner there are 1 switch (for now) and 0-6 sensors, but it will always have a `host`)

Thanks for reviewing, btw

bramkragten commented:
@azogue No, but that would be a nice addition as well! I think he means this [component](https://github.com/home-assistant/home-assistant/tree/dev/homeassistant/components/remote) for the remote control services...

azogue commented:
@pvizeli, ok, now I'm looking at the `remote` component (a `ToggleEntity` with a `send_command` method). 
I think I could change the main `switch` for a `remote` entity, and link the send raw command service to the `send_command` service of the remote, but the `sync` service doesn't have any meaning for the vacuum cleaner, and the other services would need to get a prefix-ed name, such as `xiaomi_vacuum_set_fanspeed`, isn't it? 
Don't know, I don't like very much any of the two options (switch or remote), but when I think of the remote, I see the 'stick with buttons', so maintaining the switch looked more natural to me.

About the idea of the vacuum cleaner as a platform of the remote component, instead of its own component, if I'm right (_I'm sure I'm not_), that possibility removes the integration of the associated sensors as HA entities, no?, which were the main purpose of this PR.

pvizeli commented:
That was only a question. You know it better for you platform/component if it could make sense or not :)

azogue commented:
ok @pvizeli!, I'm sorry, I'm an apprentice before anything else, so every question I take seriously ;-)
I think the schema of component with services, sensors and actuators is the better option here, but I don't know if the increasing number of components (one per 'smart' and complex device) is the desired evolution for HA.

olskar commented:
Personally I think the switch approach makes more sence. Furthermore, if there is problems implementing the remotecontrol-move of the vaccum, perhaps it should be left out. I feel like using home assistant as a remote control for the robot might be hard to achieve in a good way..?

olskar commented:
@azogue Is it also possible to implement locate service? It is supported by python-mirobo at least :)

azogue commented:
Hi @olskar, in this PR I'm only trying to implement the services proposed in #8416 by @bramkragten, but moving to a HA component. But yes, the intention is to implement all services supported by `python-mirobo`, and I hope, in a near future, even get the maps in a HA camera.

Adding more services could be done quickly in other PR, if this is merged. And adding more sensors too (consumible status, p.ex).

olskar commented:
Thanks for your work, looking forward to the merge! :)

rytilahti commented on homeassistant/components/services.yaml:
```diff
@@ -533,3 +533,88 @@ knx:
+        example: '90'
+
+      duration:
+        description: Duration of the movement?
```
A good question, honestly I don't know but that is what it is called in the payloads. Changing it doesn't seem to change the behavior though.

azogue replied:
I know, went a little crazy when testing these commands... I'm not sure about them (they where in the original PR from @bramkragten, and they are in your library, so if they are implemented, perhaps some future user will enlighten us with a proper usage.

rytilahti commented on homeassistant/components/xiaomi_vacuum.py:
```diff
@@ -0,0 +1,418 @@
+            elif self.state < 20:
+                returning_icon += '-outline'
+            return returning_icon
+        return self._icon
```
Would it make sense to have an icon also for the state? At least for when it errors or is cleaning?

azogue replied:
well, it seems like the sensors aren't very orthodox, so, we'll see how this comes out...

rytilahti commented:
Hi @azogue, your patch makes it all so much nicer (even when I don't really like the need to split those sensors & switches all around the source tree, but that's the way it is :)), so thank you a lot! I have been running this PR for a day or so without any bigger problems.

A couple of questions:
* Would it make sense to expose all the sensors when none is defined instead of other way around?
* I'm not definite about this, but shouldn't the example IP be marked as a string as required to be in the data payload?

A note about consumables and maps, introducing those would be a natural next step. I'm unsure how easy it will be to get access to the map from this robot though, all of it seems to be done remotely..

After briefly checking out your discussion on #8628, would it make sense to postpone this in favor of generic control approach? It may also make sense to generalize some sort of consumable states (and perhaps errors too?) alongside with the common commands.

@olskar reason for manual controls were added on request, as it can be useful when your robot is, e.g., parked under a bed.

azogue commented:
> Would it make sense to expose all the sensors when none is defined instead of other way around?

Was my first approach, but then I thought that way can't config 0 sensors, I can change it, what you decide...

> I'm not definite about this, but shouldn't the example IP be marked as a string as required to be in the data payload?

Could be, I need to test it better. Now I'm trying the PR with the schema:
```
vacuum:
- platform: xiaomi_vacuum
  # xiaomi config...
```
I'be commenting on it...

rytilahti commented:
Ah, indeed. I think all (or most?) sensors work the same way (you have to define them separately), so it is better to keep the consistency also here. I just don't personally like the idea of copy&pasting loads of configuration items.. (and it also seems to cause problems with some people on the forums with incorrectly formatted configs etc.)

👍 for the second part. Sounds like a sane thing to do in the long run.

azogue committed:
`icon_for_battery_level` util method

azogue committed:
Xiaomi vacuum as platform of new component vacuum

- Created new component `vacuum` from a ToggleEntity.
- Add services `turn_on`, `turn_off`, `cleaning_play_pause`, `stop`, `return_to_base`, `locate`, `set_fanspeed` and `send_command`.
- Remove the main switch for the xiaomi vacuum (the toggable main entity is the switch).
- Add `support flags` for the common services
- Assign default icons to sensors and the switch (`mdi:google-circles-group` looks like the robot!)
- Move services descriptions to a yaml file for the new component.
- Update requirements_all.
- Update coveragerc.

azogue commented:
Ok, here is the first approach to a `vacuum` component.

Based on the ToggleEntity, to substitute the switch for the new entity.
* Common services are: `turn_on`, `turn_off`, `cleaning_play_pause`, `stop`, `return_to_base`, `locate`, `set_fanspeed` and `send_command`.
* The other services (for remote control the robot) are implemented as platform dependent, so they start with `xiaomi_`: `'xiaomi_remote_control_move`, `xiaomi_remote_control_move_step`, `xiaomi_remote_control_start` and `xiaomi_remote_control_stop`.
* The associated sensors are also optional, so it's an each platform-decision, along with the map-cameras (in the neato vacuum)
* I've created the `SUPPORT_*` flags and the `supported_features` property as I think this is what the frontend uses to render the custom UI, am I wrong? (there is not a custom UI for now...). This way, there is no need to implement all common services for each botvac.

Now, the yaml config to get the xiaomi is:
```yaml
vacuum:
- platform: xiaomi_vacuum
  name: robot_aspirador
  host: !secret xiaomi_vacuum_robot_ip
  token: !secret xiaomi_vacuum_robot_token
  sensors:
    - state
    - battery
    - fanspeed
    - clean_area
    - clean_time
```
And entities created for this config sample are:
```yaml
group:
  Xiaomi Vacuum Cleaner Robot:
    - vacuum.robot_aspirador
    - sensor.robot_aspirador_state
    - sensor.robot_aspirador_battery
    - sensor.robot_aspirador_fanspeed
    - sensor.robot_aspirador_clean_area
    - sensor.robot_aspirador_clean_time
```

It should be easy to migrate the `neato` component and also adapt the new `roomba` (PR #8628) as well.

I await for your comments and criticisms...

azogue commented:
I think a PR is needed now in `homebridge-homeassistant` to expose the botvacs as switches to HomeKit (this is lost with the change, I think).

azogue committed:
fix coveragerc

pschmitt commented on homeassistant/util/icon.py:
```diff
@@ -0,0 +1,17 @@
+    if charging:
+        icon += '-charging'
+    if 20 <= battery_level < 100:
+        icon += '-{}'.format(int(battery_level / 20) * 20)
```
This does not take into account the missing `battery-charging-30`, `battery-charging-50` and `battery-charging-70` icons.
https://materialdesignicons.com/

azogue replied:
No, as I don't know why are 30/50/70 values but no 10 or 90, I'm using only the 20-multipliers... I could change this...

pschmitt replied:
I think using all the icons would be better, and we could easily add more when mdi releases more.

pschmitt commented:
Nice work, I just have a few remarks.
What's the difference between `turn_off` and `stop`?

pvizeli commented:
@azouge cool. A smart hint. The vacum should not include sensors. The Xiaomi stay a own component they discovery the vacum component and also sensors. But a core component don't discovery other components.

azogue commented:
@pschmitt, 

> What's the difference between turn_off and stop?

Turn off is stopping AND returning to the dock, stop only does that. It's a bit redundant, but I think more options are easier for the new platforms to come. You don't need to implement all methods for `VacuumEntity` for each platform, but just adjust the `support_flags`

azogue commented:
Hi @pvizeli, 

> The Xiaomi can stay a own component they discovery the vacuum component and also sensors. But a core component don't discovery other components.

Sounds a bit too convoluted, no? That way, any botvac platform with extra sugar, like the map camera in the neato (and I hope a similar one for the Xiaomi in the future), would need its own component, isn't it?
I left out the sensors of the main component `vacuum` to not force the sensors implementation for all botvacs, but I think these are pretty common. 
What do you mean with 'core component'? The xiaomi now it's a platform...

How can I implement optional sensors (and camera) for the new `VacuumEntity` without the need to create extra HA components?

azogue commented:
For the **custom UI**, I imagine a card most like the `media_player` one, but with buttons to call the common services, and at least showing 2 sensors / (or _vacuum properties_): current state and battery status. 
Making it with the ability to fetch images from a camera and showing the card as a toolbar with buttons as caption of the camera (or by itself if the camera is not supported for the platform) would be awesome.
But for all that I need guiding, help and time, or a kind soul to do it...

pvizeli commented:
You can also add "sensor" values as attribute and you can use it for template sensors. But vacuum can not make sensor platform

azogue committed:
fix battery icon helper to use more icons

azogue commented:
@pvizeli,

Well, that was the origin point, and the primary intention of the PR was to remove the need of the custom templating sensors...

Is that a 'HA philosophy' limitation? Why, if a main (but not core?) component can load other platform components, a platform of one component can't load them? Sorry, but I can't see the difference, if you could explain it to me I will be pleased.

Other option:
If the `battery_level` and the `vacuum_status`, at least, are exposed as properties for the general VacuumEntity, and then the custom UI gets them to show in the UI, I suppose the problem would be avoided, leaving any other 'rare' sensor to be exposed by the user. What do you think about it?

pschmitt commented:
I must agree with @pvizeli, having both attributes and sensors that all have the same value seems utterly redundant. This would greatly simplify the `vacuum` platform as well (not that it is that complicated, I hacked the roomba implementation in ~1h).

> Other option:
If the battery_level and the vacuum_status, at least, are exposed as properties for the general VacuumEntity, and then the custom UI gets them to show in the UI, I suppose the problem would be avoided, leaving any other 'rare' sensor to be exposed by the user. What do you think about it?

That's a good idea (unless there is a vacuum that does support reading its battery level or status - I doubt there si one). At a later point we could even expose config options for the UI, this way users could specify which sensors/properties they want to be displayed in the custom UI.

azogue commented:
> That's a good idea (unless there is a vacuum that does support reading its battery level or status - I doubt there si one). At a later point we could even expose config options for the UI, this way users could specify which sensors/properties they want to be displayed in the custom UI.

I think, but I could be wrong, that these implementation differences can be taken care of with the `supported_features` property. So perhaps that's the way to go, to implement 'sensors/attributes/properties' with an associated support flag, to render them in the UI.

> (not that it is that complicated, I hacked the roomba implementation in ~1h).
Love to hear that!

azogue committed:
remove sensors, create properties and support flags for custom UI

azogue commented:
Ok, I'm going with the custom UI, so removing sensors and making state attributes.
This is how is looking now:
![captura de pantalla 2017-07-26 a las 14 02 45](https://user-images.githubusercontent.com/9748301/28620011-73f75c0e-720b-11e7-9dc9-174234818763.png)

I need to clean, implement the switch component, and make a different PR for the UI, but I'm optimistic !

pschmitt commented:
Yeah `supported_features` should work.
By the way, the base `vacuum` entity has a bit too much services, I'd suggest moving the `locate` and the `set_fanspeed` to `xiami_vaccuum` since there does not seem to be another vacuum that supports these.

azogue committed:
cleaning

azogue committed:
updated state_attrs for filtering in UI, renamed platform to simply `xiaomi`

azogue commented:
Final changes, I hope:

* Platform renamed to `xiaomi`, as now there is no name conflict with the xiaomi sensors, **because no sensors are created in this PR** ;-)
* State attributes renamed to be filtered in the attrs table, because they are making buttons and so with the custom UI

About moving the `locate` and `set_fanspeed` to the xiaomi platform, I think it's a shame because within the generic, the custom UI gets curated info of them. 

They are **all optional**, except the `turn_on` and `turn_off` actions (the class is derived from `ToggleEntity`). So, if using the support flags correctly, it should work as they don't exist.

The final UI is this:
![captura de pantalla 2017-07-26 a las 19 52 37](https://user-images.githubusercontent.com/9748301/28636573-d7928c4c-723e-11e7-83ca-5b0d5330f143.png)

![captura de pantalla 2017-07-26 a las 19 53 11](https://user-images.githubusercontent.com/9748301/28636582-e24fd45a-723e-11e7-8000-9f1aeb53a32f.png)

So, when I propose the ha-polymer PR, I'll be waiting for your comments.
I think this is a good start for the `vacuum` component.

azogue committed:
fix platform rename

filcole commented:
I like the interface.  Here's a few comments about that specifically
thinking of the Dyson 360 Eye that can also provide:
1) a map of where it's cleaned/been
2) a location of where it's currently located.
3) indication and control of it's cleaning mode (high power vs quiet mode)

See cool stuff recently done by @CharlesBlone here
https://github.com/CharlesBlonde/libpurecoollink

On 26 July 2017 at 19:15, Eugenio Panadero <notifications@github.com> wrote:

> Final changes, I hope:
>
>    - Platform renamed to xiaomi, as now there is no name conflict with
>    the xiaomi sensors, *because no sensors are created in this PR* ;-)
>    - State attributes renamed to be filtered in the attrs table, because
>    they are making buttons and so with the custom UI
>
> About moving the locate and set_fanspeed to the xiaomi platform, I think
> it's a shame because within the generic, the custom UI gets curated info of
> them.
>
> They are *all optional*, except the turn_on and turn_off actions (the
> class is derived from ToggleEntity). So, if using the support flags
> correctly, it should work as they don't exist.
>
> The final UI is this:
> [image: captura de pantalla 2017-07-26 a las 19 52 37]
> <https://user-images.githubusercontent.com/9748301/28636573-d7928c4c-723e-11e7-83ca-5b0d5330f143.png>
>
> [image: captura de pantalla 2017-07-26 a las 19 53 11]
> <https://user-images.githubusercontent.com/9748301/28636582-e24fd45a-723e-11e7-8000-9f1aeb53a32f.png>
>
> So, when I propose the ha-polymer PR, I'll be waiting for your comments.
> I think this is a good start for the vacuum component.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/home-assistant/home-assistant/pull/8623#issuecomment-318138136>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AFy_vhRIGMNkSM3GUKLQqJaq5gNXwn9Tks5sR4JSgaJpZM4Ogoc4>
> .
>



-- 
Phil

pschmitt commented:
At least some[1] of the roombas expose their position and cleaning mode
settings as well. Once I wrap my head around it mapping is possible as well.

[1] https://github.com/koalazak/dorita980/issues/48

On Thu, Jul 27, 2017, 02:19 Phil Cole <notifications@github.com> wrote:

> I like the interface. Here's a few comments about that specifically
> thinking of the Dyson 360 Eye that can also provide:
> 1) a map of where it's cleaned/been
> 2) a location of where it's currently located.
> 3) indication and control of it's cleaning mode (high power vs quiet mode)
>
> See cool stuff recently done by @CharlesBlone here
> https://github.com/CharlesBlonde/libpurecoollink
>
> On 26 July 2017 at 19:15, Eugenio Panadero <notifications@github.com>
> wrote:
>
> > Final changes, I hope:
> >
> > - Platform renamed to xiaomi, as now there is no name conflict with
> > the xiaomi sensors, *because no sensors are created in this PR* ;-)
> > - State attributes renamed to be filtered in the attrs table, because
> > they are making buttons and so with the custom UI
> >
> > About moving the locate and set_fanspeed to the xiaomi platform, I think
> > it's a shame because within the generic, the custom UI gets curated info
> of
> > them.
> >
> > They are *all optional*, except the turn_on and turn_off actions (the
> > class is derived from ToggleEntity). So, if using the support flags
> > correctly, it should work as they don't exist.
> >
> > The final UI is this:
> > [image: captura de pantalla 2017-07-26 a las 19 52 37]
> > <
> https://user-images.githubusercontent.com/9748301/28636573-d7928c4c-723e-11e7-83ca-5b0d5330f143.png
> >
> >
> > [image: captura de pantalla 2017-07-26 a las 19 53 11]
> > <
> https://user-images.githubusercontent.com/9748301/28636582-e24fd45a-723e-11e7-8000-9f1aeb53a32f.png
> >
> >
> > So, when I propose the ha-polymer PR, I'll be waiting for your comments.
> > I think this is a good start for the vacuum component.
> >
> > —
> > You are receiving this because you are subscribed to this thread.
> > Reply to this email directly, view it on GitHub
> > <
> https://github.com/home-assistant/home-assistant/pull/8623#issuecomment-318138136
> >,
> > or mute the thread
> > <
> https://github.com/notifications/unsubscribe-auth/AFy_vhRIGMNkSM3GUKLQqJaq5gNXwn9Tks5sR4JSgaJpZM4Ogoc4
> >
> > .
> >
>
>
>
> --
> Phil
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/home-assistant/home-assistant/pull/8623#issuecomment-318220460>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AACT_tAL2CNU3JK5r7jSxNyMI9B7dL1xks5sR9eLgaJpZM4Ogoc4>
> .
>
-- 

Philipp Schmitt

pschmitt commented:
@azogue Using your latest code, the services `locate` and `set_fanspeed` are still registered despite not having set `SUPPORT_FANSPEED` or `SUPPORT_LOCATE`.

Here's my code: https://gist.github.com/pschmitt/eef8b01e051223905d4368598426b671

azogue commented:
I'm thinking about changing the `set_fanspeed` service to accept a string (instead of a positive int) and make a **`fanspeed_list` property**, like the `fan` component, to get an input select in the more-info card, and be able to work with speed 'steps' (low, medium, high and so). This is the way the majority of botvacs are working, as I see. 

For the Xiaomi, which can set the fanspeed with any speed (0-100), I will use the steps the official app is using for the `fanspeed_list`, and also use the `set_fanspeed` service to accept any number, so other platforms can define its speed steps easily and no feature is lost, **how do you see that?**

About the position maps, @filcole, the intention is to integrate them in the component, so the `neato` could adapt all its features, and also others botvacs, like the Dyson one. But that would arrive in other PR, because now I don't have much time to investigate and we don't know yet how to get the Xiaomi maps from the server, so I couldn't test it.

If this PR merges, I encourage you to implement other botvac platforms over it, making the necessary changes to get the maps and more ;-)

azogue commented:
Hi @pschmitt,
> Wouldn't it make more sense to only register services if any (at least one) of the vacuums supports it?

I don't know how to do that in a clean way, nor I know if it is an 'orthodox' way of doing things, for this we need @pvizeli to shed some light on the subject...

The code looks great, btw, but you need to remove the `async_dispatcher_send(self.hass, SIGNAL_UPDATE_DATA)` call, there are no more entities to update now.

pschmitt commented:
@azogue Thanks, I updated my [Gist](https://gist.github.com/pschmitt/eef8b01e051223905d4368598426b671).

I'm fine with some services that actually do nothing, I believe that's how `media_player` works as well.

azogue committed:
change fanspeed and expose `fanspeed_list` to use speed steps

azogue commented:
Ok, final question, I hope,

**Do I need to change the `state` of the vacuum entity?** 

Now it's a `STATE_ON` / `STATE_OFF` device (as `ToggleEntity`), so the default UI shows a switch. The custom UI also shows the switch to call the `turn_on` / `turn_off` services, so the question is independent from the UI. But I don't know **if the `state` has to be the `status` of the device**. 

Now the status is an optional string property linked to the `SUPPORT_STATUS` flag, but if all botvacs have one, it could be a basic prop.
In that thought, the `SUPPORT_TURN_ON` and `SUPPORT_TURN_OFF` flags have no reason to be, isn't it?

I'm thinking about automations and so on. In any case, the status is accessible (`{{states.vacuum.robot_aspirador.attributes.vacuum_status}}`), but one way is more direct.

rytilahti commented:
I'm very glad that you included the possibility to change the fanspeed, just wondering if it should be named just `mode` instead of `fanspeed` though?

Also, considering the sensors are not anymore exposed, what would be the preferred way to signal the user about an error? I feel that parsing the status string is not a very nice solution, do you have a strong preference for not exposing an error state explicitly, even as an attribute? This would allow using `robot.attributes.has_error` (or similarly worded attribute) for automations in a generic manner.

:+1: for not exposing the raw interface on xiaomi. According to some user reports the official app will get confused if it receives unknown fan speeds.

All in all this looks really great and I hope it will get merged soon!

azogue commented:
> Also, considering the sensors are not anymore exposed, what would be the preferred way to signal the user about an error? I feel that parsing the status string is not a very nice solution

The 'Error' attribute is present for the Xiaomi in its state_attrs, so I suppose the official way of getting that is creating a `template sensor` with that attribute or the status one. My question about the entity `state` is about this, thinking in the automation problem.

> 👍 for not exposing the raw interface on xiaomi. According to some user reports the official app will get confused if it receives unknown fan speeds.

Is not what I did. The list is in the UI, to select the default modes, but you can call the service with any speed (integer), along with the keys. It's better!

About the official app, I saw what you mean when testing this PR, is not a problem. If the speed is not any of the default list, it shows nothing, but you can press it to cycle though the default labels, and worked perfectly when other speed was settled.

azogue commented:
@rytilahti, if you want to test the UI, I'm using it now in the 'production' machine as [`custom_component`](https://github.com/azogue/hass_config/tree/master/custom_components/vacuum), using the `custom_ui` option with this:
https://github.com/azogue/hass_config/blob/master/www/custom_ui/state-card-custom_vacuum.html

The yaml config to get all working:

```yaml
homeassistant:
  customize:
    vacuum.robot_aspirador:
      custom_ui_state_card: custom_vacuum

vacuum:
- platform: xiaomi
  name: robot_aspirador
  host: !secret xiaomi_vacuum_robot_ip
  token: !secret xiaomi_vacuum_robot_token

group:
  Xiaomi Vacuum Robot:
    control: hidden
    entities:
      - vacuum.robot_aspirador
```

rytilahti commented:
> The 'Error' attribute is present for the Xiaomi in its state_attrs, so I suppose the official way of getting that is creating a template sensor with that attribute or the status one. My question about the entity state is about this, thinking in the automation problem.

Yes, although it says currently `No error` instead of being empty, requiring parsing when doing automations? As I am not very familiar with the automations, maybe someone chimes in, although would leaving it completely empty when there is no error work better?

> Is not what I did. The list is in the UI, to select the default modes, but you can call the service with any speed (integer), along with the keys. It's better!

Indeed! It was a bad formulation from my side, I meant that giving a list of options instead of having a way to choose between 1-100 is so much better. Also good to know that the "problem" with the official app is not so bad as I understood.

As I was for the first time ever to get the development environment for the interface working, I tested the PR directly there. Here's how it currently looks after your newest updates, it's simply gorgeous in my opinion :)

![image](https://user-images.githubusercontent.com/3705853/28682014-c318ba8a-72fb-11e7-86f4-e7488cca112d.png)

Btw, do you have any plans how to introduce the consumable states or the global statistics (total area cleaned, total cleaning time, ..)? These are probably xiaomi-specific and should be just added as state attributes.

Also as a side-note, I'm not sure if exposing the "do not disturb" state is very useful, but besides that it is reported in the status always as "on". If this is really wanted, `dnd_status()` should be used for checking its state, it also returns the time-range when it is active if that's the case.

olskar commented:
So nice looking! Cant wait to see this merged, my floors needs cleaning :)

rytilahti commented on homeassistant/components/vacuum/services.yaml:
```diff
@@ -0,0 +1,131 @@
+      example: 'vacuum.xiaomi_vacuum_cleaner'
+
+cleaning_play_pause:
+  description: Pause the cleaning task or replay it.
```
replay a task?

azogue replied:
sorry, `resume` better, isn't it?

rytilahti replied:
It sounds better yes, although I suppose it also starts the cleaning, if it's not cleaning already?

azogue replied:
exactly

rytilahti commented on homeassistant/components/vacuum/services.yaml:
```diff
@@ -0,0 +1,131 @@
+      example: 'vacuum.xiaomi_vacuum_cleaner'
+
+send_command:
+  description: Send 'raw' command to the vacuum cleaner.
```
Nitpick, but "a raw command" maybe?

azogue replied:
ok

rytilahti commented on homeassistant/components/vacuum/xiaomi.py:
```diff
@@ -0,0 +1,375 @@
+SPEED_MAX = ('Max', 90)
+FANSPEED_LABELS, FANSPEED_VALUES = list(
+    zip(*[SPEED_LOW, SPEED_MEDIUM, SPEED_HIGH, SPEED_MAX]))
+
```
Would this be better as a enum or a dict?
```
class Speed(enum.Enum):
    Quiet = 38
    Balanced = 60
```

azogue replied:
_I'm quitting from enums_, after a decade working with C ;-)

Seriously, I've seen them in very very few places in Python, and I think is for the better, but we can wait to see what _the elders_ have to say about it

rytilahti replied:
Agreed. I think there is some resistance among the core developers to use such instead of plain old globals. But can't say I didn't try ;-)

rytilahti commented on homeassistant/components/vacuum/xiaomi.py:
```diff
@@ -0,0 +1,375 @@
+    SERVICE_MOVE_REMOTE_CONTROL_STEP: {
+        'method': 'async_move_remote_control_step',
+        'schema': SERVICE_SCHEMA_REMOTE_CONTROL},
+}
```
Considering the keys here are used only in this place, in my opinion it would enhance readability to not to define these above as globals but just use them directly here.. Opinions regarding to this may vary though..

azogue replied:
that's the way it's done everywhere in HA, as I know

rytilahti replied:
Indeed, fair enough.

rytilahti commented on homeassistant/components/vacuum/xiaomi.py:
```diff
@@ -0,0 +1,375 @@
+        """Return the status of the vacuum cleaner."""
+        if self.vacuum_state is not None:
+            return self.vacuum_state.state
+        return None
```
Returning `None` explicitly is not necessary, but as you wish.

rytilahti commented on homeassistant/components/vacuum/xiaomi.py:
```diff
@@ -0,0 +1,375 @@
+    @property
+    def fanspeed_list(self):
+        """Get the list of available fan speed steps of the vacuum cleaner."""
+        return list(FANSPEED_LABELS)
```
Even when not using dict/enum, the labels are already a list so no need to reconstruct it here.

azogue replied:
I think they are not, they are a tuple.

rytilahti replied:
Oops, indeed. Maybe one of the reasons I do really like a more pythonic approach to keep it more readable ;-) Anyways, not for me to decide, let's see what others have to say.

rytilahti commented on homeassistant/components/vacuum/__init__.py:
```diff
@@ -0,0 +1,287 @@
+        if self.fanspeed is not None:
+            data['vacuum_fanspeed'] = self.fanspeed
+            data['vacuum_fanspeed_list'] = self.fanspeed_list
+
```
Btw, did you consider my earlier proposal about renaming fanspeed to mode?

azogue replied:
The reason for the prefix is [this](https://github.com/home-assistant/home-assistant-polymer/blob/master/src/util/hass-util.html#L37). To proper identify and filter specific attributes from render them in the attributes table. I don't want to hide some battery level for any other entity over there... 

I suppose that's not the better approach, but I'm waiting @pvizeli to comment on this.

rytilahti replied:
I'm unsure how to interpret that piece of code, is it that those are hidden from the attributes lists per default?

A nice thing with all of batteries reporting `battery_level`s would make it easy to have a component/automation (e.g. https://community.home-assistant.io/t/script-to-track-all-devices-with-a-battery-level/2596) for checking the battery levels using that attribute, without having to have a list of different matchers for that.

azogue replied:
not hidden but filtered out when rendering the state attributes table in the more-info view. 
You can access them in `{{states.vacuum.robot_aspirador.attributes.vacuum_status}}` and so.

The extra complexity is because of the filter, to not get false positives.

azogue commented on homeassistant/components/vacuum/__init__.py:
```diff
@@ -0,0 +1,287 @@
+        return self.hass.async_add_job(partial(self.set_fanspeed, **kwargs))
+
+    def cleaning_play_pause(self, **kwargs):
+        """Pause the cleaning task or replay it."""
```
ok

rytilahti replied:
So I'm not sure if start_pause is nice either, but `play` just sounds plainly wrong to my ear :-)

azogue replied:
I think I'll go with `start_pause`, stripping also the 'cleaning_' prefix (that was there from copy-pasting the media player...)

azogue commented:
Hi @rytilahti, 

> Btw, do you have any plans how to introduce the consumable states or the global statistics (total area cleaned, total cleaning time, ..)? These are probably xiaomi-specific and should be just added as state attributes.

I'm exposing the attributes the `VacuumStatus` class has, if more attributes are needed, I think first we need to expose them in that results wrapper, and then add them here.

> Also as a side-note, I'm not sure if exposing the "do not disturb" state is very useful, but besides that it is reported in the status always as "on". If this is really wanted, dnd_status() should be used for checking its state, it also returns the time-range when it is active if that's the case.

For me, the dnd attribute is 'off'. In the official app you can turn it on and off as I recall.

As the _error attribute_, I removed the `"Error": "No error"` attribute, so it's added only when there is an error. So, for automations (I'm thinking of error notifications), we can ask if `{{ 'Error' in states.vacuum.robot_aspirador.attributes }}` to check. How do you see it?

azogue commented:
My automation for error notification for the Xiaomi:
```yaml
automation:
- alias: Botvac error notification
  trigger:
    platform: state
    entity_id: vacuum.robot_aspirador
  condition:
    condition: template
    value_template: '{{ "Error" in states.vacuum.robot_aspirador.attributes }}'
  action:
    service: telegram_bot.send_message
    data_template:
      title: '*Botvac Error*'
      message: >
        El robot aspirador ha fallado: _{{ states.vacuum.robot_aspirador.attributes.Error }}_
        (Status: {{ states.vacuum.robot_aspirador.attributes.vacuum_status }}, batt: {{ states.vacuum.robot_aspirador.attributes.vacuum_battery }} %)
      target: '{{ states.sensor.telegram_default_chatid.state | int }}'
      disable_notification: false
      inline_keyboard:
        - Locate botvac:/service_call vacuum.locate vacuum.robot_aspirador, To dock:/service_call vacuum.return_to_home vacuum.robot_aspirador
        - Start:/service_call vacuum.start_pause vacuum.robot_aspirador, Stop:/service_call vacuum.stop vacuum.robot_aspirador, +:/init
```

* The `target` is variable because it's targeted to the people at home, so can be me, my couple, or both (different Telegram chats)
* The inline keyboard, with an AppDaemon app that handles the Bot, allows easy making of buttons with associated actions. In this case, service calls to the vacuum component.

azogue committed:
minor fixes

- Rename service `start_pause`
- Add 'Error' attribute only if `got_error`.
- Minor changes

pschmitt commented:
I guess the error attribute is something all vacs should have. It'd make sense to make it a `vacuum` attribute, like `battery_level` or `status`, wouldn't it?

azogue commented:
> I guess the error attribute is something all vacs should have. It'd make sense to make it a vacuum attribute, like battery_level or status, wouldn't it?

Don't know. As I see, what you're requesting has to be an **`Entity` property**, as now the `unavailable` state is, so I think that can be addressed in other PR (I need to finish this!!)

azogue commented:
@fabaff, @pvizeli, the labels of this PR are wrong (many things have changed):

`component: xiaomi_vacuum`  -->  `component: vacuum`
`platform: sensor.xiaomi_vacuum` needs to be removed and `platform: vacuum.xiaomi` to be added.

rytilahti commented:
Hi @azogue 

> I'm exposing the attributes the VacuumStatus class has, if more attributes are needed, I think first we need to expose them in that results wrapper, and then add them here.

I'm generally against introducing more implicit calls inside the backend library as we already do the handshaking & the corresponding command for every call, requesting more information (scheduling information, consumables, dnd state,  when necessary is the responsibility of the library user.

> For me, the dnd attribute is 'off'. In the official app you can turn it on and off as I recall.

Uh, okay. I just played around with the official app and the setting. Looks like the official app does something besides just what python-mirobo does when calling `mirobo dnd on 08 03 21 00`, so if one is not using the official app to set that setting its value is inconsistent.

You can check this easily by modifying the settings and checking after changes with `mirobo -dd status` and look into the raw JSON. 

> As the error attribute, I removed the "Error": "No error" attribute, so it's added only when there is an error. So, for automations (I'm thinking of error notifications), we can ask if {{ 'Error' in states.vacuum.robot_aspirador.attributes }} to check. How do you see it?

That sounds fine for me, someone who knows about automations will maybe chime in, I'm just using appdaemon app for the controls so it's not a problem for me, I just happen to like clean interfaces.

P.S., `component: xiaomi_vacuum` should probably also be there, as this involves obsoleting and removing it, agreed otherwise :-)

P.P.S., xiaomi also supports "spot cleaning", adding a service call for that would not be very hard, which could be useful in combination with the manual controls. However, that can also be implemented later if someone asks for it...

azogue committed:
rename state attrs

MartinHjelmare commented on homeassistant/components/vacuum/__init__.py:
```diff
@@ -0,0 +1,287 @@
+        data = {}
+
+        if self.status is not None:
+            data['Status'] = self.status
```
Attributes should be all lower case.

azogue replied:
ok, I'll change that (also in the polymer PR for the custom UI)

MartinHjelmare commented on homeassistant/components/vacuum/__init__.py:
```diff
@@ -0,0 +1,287 @@
+            data['Status'] = self.status
+
+        if self.battery_level is not None:
+            data['Battery'] = self.battery_level
```
Import and use `ATTR_BATTERY_LEVEL` from const.py.

azogue committed:
rename state attrs

MartinHjelmare commented on homeassistant/components/vacuum/services.yaml:
```diff
@@ -0,0 +1,131 @@
+turn_on:
+  description: Starts a new cleaning task.
```
"Start a new cleaning task."

MartinHjelmare commented on homeassistant/components/vacuum/services.yaml:
```diff
@@ -0,0 +1,131 @@
+      example: 'vacuum.xiaomi_vacuum_cleaner'
+
+return_to_base:
+  description: Tells the vacuum cleaner to return to its dock.
```
"Tells the vacuum cleaner to return to its dock."

azogue replied:
sorry, It sounds to me like "the service" _does_ something..., I'll try to remember this because I do that a lot. Thanks!

MartinHjelmare commented on homeassistant/components/vacuum/services.yaml:
```diff
@@ -0,0 +1,131 @@
+      description: Parameters for the command.
+      example: '[22,0,6,0]'
+
+set_fanspeed:
```
`set_fan_speed:`

MartinHjelmare commented on homeassistant/components/vacuum/services.yaml:
```diff
@@ -0,0 +1,131 @@
+      example: '[22,0,6,0]'
+
+set_fanspeed:
+  description: Set the fanspeed of the vacuum cleaner.
```
"Set the fan speed of the vacuum cleaner."

MartinHjelmare commented on homeassistant/components/vacuum/services.yaml:
```diff
@@ -0,0 +1,131 @@
+      description: Name of the botvac entity.
+      example: 'vacuum.xiaomi_vacuum_cleaner'
+
+    fanspeed:
```
`fan_speed:`

MartinHjelmare commented on homeassistant/components/vacuum/services.yaml:
```diff
@@ -0,0 +1,131 @@
+
+    duration:
+      description: Duration of the movement?
+      example: '1500'
```
No newline at end of file.

MartinHjelmare commented on homeassistant/components/vacuum/xiaomi.py:
```diff
@@ -0,0 +1,373 @@
+Support for the Xiaomi vacuum cleaner robot.
+
+For more details about this platform, please refer to the documentation
+https://home-assistant.io/components/vacuum.xiaomi_vacuum/
```
https://home-assistant.io/components/vacuum.xiaomi/

azogue replied:
oops, didn't change that with the change of name

MartinHjelmare commented on homeassistant/components/vacuum/xiaomi.py:
```diff
@@ -0,0 +1,373 @@
+
+from homeassistant.components.vacuum import (
+    VacuumDevice, DOMAIN,
+    DEFAULT_ICON, PLATFORM_SCHEMA, VACUUM_SERVICE_SCHEMA,
```
:abcd:

MartinHjelmare commented on homeassistant/components/vacuum/__init__.py:
```diff
@@ -0,0 +1,287 @@
+
+from homeassistant.config import load_yaml_config_file
+from homeassistant.const import (
+    ATTR_ENTITY_ID, ATTR_BATTERY_LEVEL, STATE_UNKNOWN,
```
:abcd:

azogue replied:
ok!

MartinHjelmare commented on homeassistant/components/vacuum/xiaomi.py:
```diff
@@ -0,0 +1,373 @@
+    SUPPORT_FANSPEED, SUPPORT_SENDCOMMAND)
+from homeassistant.config import load_yaml_config_file
+from homeassistant.const import (
+    STATE_ON, STATE_OFF, ATTR_ENTITY_ID, CONF_NAME, CONF_HOST, CONF_TOKEN)
```
:abcd:

MartinHjelmare commented on homeassistant/components/vacuum/xiaomi.py:
```diff
@@ -0,0 +1,373 @@
+SERVICE_START_REMOTE_CONTROL = 'xiaomi_remote_control_start'
+SERVICE_STOP_REMOTE_CONTROL = 'xiaomi_remote_control_stop'
+
+FANSPEEDS = {
```
`FAN_SPEEDS`

MartinHjelmare commented on homeassistant/components/vacuum/xiaomi.py:
```diff
@@ -0,0 +1,373 @@
+
+SERVICE_SCHEMA_REMOTE_CONTROL = VACUUM_SERVICE_SCHEMA.extend({
+    vol.Optional(ATTR_RC_VELOCITY):
+        vol.All(vol.Coerce(float), vol.Range(min=-0.29, max=0.29)),
```
An alternative to `vol.Range` could be `vol.Clamp` if we want to avoid throwing an error. Of course this will not be as informative to the user as an error.

azogue replied:
changed

MartinHjelmare commented on homeassistant/components/vacuum/xiaomi.py:
```diff
@@ -0,0 +1,373 @@
+    @asyncio.coroutine
+    def async_service_handler(service):
+        """Map services to methods on MiroboVacuum."""
+        _LOGGER.debug('XIAOMI async_service_handler -> %s, %s',
```
You can remove this.

azogue replied:
old debugging code

MartinHjelmare commented on homeassistant/components/vacuum/xiaomi.py:
```diff
@@ -0,0 +1,373 @@
+            yield from getattr(vacuum, method['method'])(**params)
+
+        for vacuum in target_vacuums:
+            if vacuum.should_poll:
```
Isn't this always true for this platform?

azogue replied:
yes ;-)

MartinHjelmare commented on homeassistant/components/vacuum/xiaomi.py:
```diff
@@ -0,0 +1,373 @@
+class MiroboVacuum(VacuumDevice):
+    """Representation of a Xiaomi Vacuum cleaner robot."""
+
+    def __init__(self, hass, name, host, token):
```
You don't need to pass in `hass`, since it will be set on the entity when the entity is added to the component.

azogue replied:
oops, I think that is there from when the class was a simple handler, fixing it!

MartinHjelmare commented on homeassistant/components/vacuum/xiaomi.py:
```diff
@@ -0,0 +1,373 @@
+        return self._icon
+
+    @property
+    def state(self) -> str:
```
You should not overwrite `state` but `is_on` property.

azogue replied:
removed

MartinHjelmare commented on homeassistant/components/vacuum/xiaomi.py:
```diff
@@ -0,0 +1,373 @@
+            return self.vacuum_state.battery
+
+    @property
+    def fanspeed(self):
```
`fan_speed`

MartinHjelmare commented on homeassistant/components/vacuum/xiaomi.py:
```diff
@@ -0,0 +1,373 @@
+            return speed
+
+    @property
+    def fanspeed_list(self):
```
`fan_speed_list`

MartinHjelmare commented on homeassistant/components/vacuum/xiaomi.py:
```diff
@@ -0,0 +1,373 @@
+        """Return the specific state attributes of this vacuum cleaner."""
+        if self.vacuum_state is not None:
+            attrs = {
+                'Do not disturb':
```
The common thing in home assistant it to have the attribute keys be all lower case and snake case.

azogue replied:
Yes, but the `ha-attributes` doesn't capitalise them when rendering the attrs table, I think, I'll try again.

MartinHjelmare commented on homeassistant/components/vacuum/xiaomi.py:
```diff
@@ -0,0 +1,373 @@
+            from mirobo import Vacuum
+            _LOGGER.info("Initializing with host %s (token %s...)",
+                         self._host, self._token[:5])
+            self._vacuum = Vacuum(self._host, self._token)
```
If this does I/O it should be moved to a method eg `async_update`. Doing this type of thing in a property is not common anyhow in home assistant. Since `async_update` will be called immediately after adding the entity, I suggest moving it there either way, if there's a reason not to put this in `__init__` other than I/O.

azogue replied:
this is from the old switch code, I'll review it, but I think it's a lazy wrapper, so it doesn't do I/O there.

azogue replied:
I'm removing the property accessor and moving the creation of the `Vacuum` handler to setup. It appears it doesn't do I/O at creation.

MartinHjelmare commented on homeassistant/components/vacuum/xiaomi.py:
```diff
@@ -0,0 +1,373 @@
+            "Unable to stop: %s", self.vacuum.stop)
+
+    @asyncio.coroutine
+    def async_set_fanspeed(self, fanspeed: str, **kwargs):
```
```py
def async_set_fan_speed(self, fan_speed: str, **kwargs):
```
Consider dropping the annotation on the positional argument since the annotation for the function is not complete, ie no return value annotation, and for consistency.

azogue replied:
I'm going to drop it... that was copy-pasted from the old approach.

MartinHjelmare commented on homeassistant/components/vacuum/xiaomi.py:
```diff
@@ -0,0 +1,373 @@
+            try:
+                fanspeed = int(fanspeed)
+            except ValueError as exc:
+                _LOGGER.error("Fan speed step not recognized (%s). "
```
I think we should consider having a fixed set of fan speeds in the home assistant base component service interface, do the validation in the service schema, and do the conversion to platform specific values in the platform. That way we'll avoid errors and have more control.

azogue replied:
It's not clear, each botvac has a number of speeds or 'modes', and they're labelled distinctly in its respective official app, these are not the 3 speeds of a normal fan, I think.

I think that can be revisited when three or more botvacs are implemented, the roomba is waiting for this.

MartinHjelmare reviewed (changes requested):
Very cool!

All base component services needs a corresponding module level function that calls the service. See eg the cover component.

A new vacuum demo platform needs to be added and this should be used to write tests for the base component.

Note my [comment](https://github.com/home-assistant/home-assistant/pull/8623#discussion_r130239224) about having a common fan speed service interface in home assistant and doing the conversion in the platforms. I think that's a better option than trying to validate the service data in the entity methods.

MartinHjelmare commented on homeassistant/components/vacuum/xiaomi.py:
```diff
@@ -0,0 +1,373 @@
+
+    @asyncio.coroutine
+    def async_set_fanspeed(self, fanspeed: str, **kwargs):
+        """Set the fanspeed."""
```
"Set fan speed."

azogue commented on homeassistant/components/vacuum/__init__.py:
```diff
@@ -0,0 +1,287 @@
+
+GROUP_NAME_ALL_VACUUMS = 'all vacuum cleaners'
+
+ATTR_COMMAND = 'command'
```
in this same PR?

MartinHjelmare replied:
I'm ok with that, but you decide.

azogue commented on homeassistant/components/vacuum/__init__.py:
```diff
@@ -0,0 +1,287 @@
+GROUP_NAME_ALL_VACUUMS = 'all vacuum cleaners'
+
+ATTR_COMMAND = 'command'
+ATTR_FANSPEED = 'fanspeed'
```
ok, I'll change all of them...

azogue commented on homeassistant/components/vacuum/__init__.py:
```diff
@@ -0,0 +1,287 @@
+
+        target_vacuums = component.async_extract_from_service(service)
+        params = dict(service.data)
+        _LOGGER.debug('Service call to %s: %s (%s)',
```
old debugging code, sorry

azogue commented on homeassistant/components/vacuum/__init__.py:
```diff
@@ -0,0 +1,287 @@
+        return 0
+
+    @property
+    def state(self):
```
ok, didn't know that, thanks

azogue commented on homeassistant/components/vacuum/__init__.py:
```diff
@@ -0,0 +1,287 @@
+        data = {}
+
+        if self.status is not None:
+            data['status'] = self.status
```
ok

azogue replied:
all declared now

azogue committed:
review changes: cut fan__speed, style changes, remove logging, and more

azogue committed:
add ATTR_COMMAND = 'command' to const

azogue commented on homeassistant/components/vacuum/__init__.py:
```diff
@@ -0,0 +1,287 @@
+            return
+
+        target_vacuums = component.async_extract_from_service(service)
+        params = dict(service.data)
```
sorry, I thought that `async_extract_from_service` was doing that. Fixed, and also getting the params as the `cover` does:
```python
params = service.data.copy()
params.pop(ATTR_ENTITY_ID, None)
```

azogue committed:
pop entity_id from service data

azogue committed:
remove property accessor for vacuum object

azogue committed:
lint fix

azogue committed:
fix extra attrs names

azogue committed:
module level functions for calling the services

MartinHjelmare commented on homeassistant/components/vacuum/__init__.py:
```diff
@@ -0,0 +1,360 @@
+
+
+@bind_hass
+def send_command(hass, command, params, entity_id=None):
```
Since `params` is optional in the service schema, it should be a keyword argument and default to None here.

azogue replied:
right, sorry, I'll fix it

azogue committed:
params as optional keyword for `send_command`

azogue committed:
params as optional keyword for `send_command`, remove debug logs

MartinHjelmare commented on homeassistant/components/vacuum/__init__.py:
```diff
@@ -0,0 +1,361 @@
+        """
+        return self.hass.async_add_job(partial(self.locate, **kwargs))
+
+    def set_fan_speed(self, **kwargs):
```
The methods in the base class should have the same signature as the overwritten methods in the child classes. It's good to break out the specific arguments as you've done in the xiaomi platform.
```py
def set_fan_speed(self, fan_speed, **kwargs):
```

azogue replied:
I was seeing this now, too (I'm trying to make the `demo` platform!)
Too many changes in this PR since the beginning... I'm on it

MartinHjelmare replied:
Great job so far! :+1:

MartinHjelmare commented on homeassistant/components/vacuum/__init__.py:
```diff
@@ -0,0 +1,361 @@
+        """Set fan speed."""
+        raise NotImplementedError()
+
+    def async_set_fan_speed(self, **kwargs):
```
See comment above.

MartinHjelmare commented on homeassistant/components/vacuum/__init__.py:
```diff
@@ -0,0 +1,361 @@
+        return self.hass.async_add_job(
+            partial(self.start_pause, **kwargs))
+
+    def send_command(self, **kwargs):
```
See comment above.

MartinHjelmare commented on homeassistant/components/vacuum/__init__.py:
```diff
@@ -0,0 +1,361 @@
+        """Send a command to a vacuum cleaner."""
+        raise NotImplementedError()
+
+    def async_send_command(self, **kwargs):
```
See comment above.

azogue committed:
explicit parameters for `set_fan_speed` and `send_command`

azogue committed:
Demo platform for the vacuum component

azogue committed:
vacuum tests for the Demo platform

azogue commented:
@MartinHjelmare, 

> A new vacuum demo platform needs to be added and this should be used to write tests for the base component.

It's done, can you review it?

MartinHjelmare commented on homeassistant/components/vacuum/demo.py:
```diff
@@ -0,0 +1,179 @@
+
+    def return_to_base(self, **kwargs):
+        """Tell the vacuum to return to its dock."""
+        if self.supported_features & SUPPORT_RETURN_HOME != 0:
```
See above comment.

MartinHjelmare commented on homeassistant/components/vacuum/demo.py:
```diff
@@ -0,0 +1,179 @@
+
+    def send_command(self, command, params=None, **kwargs):
+        """Send a command to the vacuum."""
+        if self.supported_features & SUPPORT_SEND_COMMAND != 0:
```
See above comment.

MartinHjelmare commented on homeassistant/components/vacuum/demo.py:
```diff
@@ -0,0 +1,179 @@
+    def send_command(self, command, params=None, **kwargs):
+        """Send a command to the vacuum."""
+        if self.supported_features & SUPPORT_SEND_COMMAND != 0:
+            command = kwargs[ATTR_COMMAND]
```
You already have `command` as an argument.

MartinHjelmare commented on tests/components/vacuum/test_demo.py:
```diff
@@ -0,0 +1,189 @@
+"""The tests for the Demo vacuum platform."""
+import unittest
+
+from homeassistant.setup import setup_component
```
:abcd:

MartinHjelmare commented on tests/components/vacuum/test_demo.py:
```diff
@@ -0,0 +1,189 @@
+        params = {"rotate": 150, "speed": 20}
+        vacuum.send_command(
+            self.hass, entity_id=ENTITY_VACUUM_BASIC,
+            command='test_command', params=params)
```
`command` is not a keyword argument for the `send_command` function. I'm surprised that this works!?

azogue replied:
don't know if that works for all python 3 versions, I'll change it

MartinHjelmare replied:
I've tested it in python 3.4.3 and it does work. :stuck_out_tongue_closed_eyes: Good to know. Maybe just for clarity, change it.

MartinHjelmare reviewed (changes requested):
You need to update `.coveragerc` and make sure that the demo platform and base component are not excluded from the test run.

azogue commented on homeassistant/components/vacuum/demo.py:
```diff
@@ -0,0 +1,179 @@
+
+    def turn_on(self, **kwargs):
+        """Turn the vacuum on."""
+        if self.supported_features & SUPPORT_TURN_ON != 0:
```
ok

azogue commented on homeassistant/components/vacuum/demo.py:
```diff
@@ -0,0 +1,179 @@
+    @property
+    def device_state_attributes(self):
+        """Return device state attributes."""
+        return {'cleaned_area': round(self._cleaned_area, 2)}
```
done

azogue committed:
some fixes

azogue committed:
don't omit vacuum

azogue committed:
vacuum tests for the Xiaomi platform

azogue committed:
fix test

azogue committed:
fix

MartinHjelmare commented on tests/components/vacuum/test_xiaomi.py:
```diff
@@ -0,0 +1,161 @@
+def mock_mirobo():
+    """Mock mock_mirobo."""
+    mock_vacuum = mock.MagicMock()
+    type(mock_vacuum.Vacuum().status()).data = {'test': 'raw'}
```
Why the use of `type`?

azogue replied:
I have very little experience with tests, mocking and so on in Python, that was 'the way' it works as I want, but, please, _enlighten me_.

MartinHjelmare replied:
I tried the line without the `type` call and didn't see a difference in the result, so I wasn't sure what the purpose was. The default for `MagicMock` is to return a new `MagicMock` when accessing any attribute, and the default return value for a `MagicMock` is also a new `MagicMock`.

Maybe I'm missing something in the bigger picture?

azogue replied:
I see... sorry, I'll change it.

I think in my first attempts was trying `mock_vacuum.Vacuum.status.XXX`, which wasn't working, and don't know how I ended up there... (but it works! ¿?)

I needed to create a pseudo-state for the vacuum to test the entity state props, because I was getting errors with `MagicMock` objects when accessing parameters. I tried until I've got it working...

MartinHjelmare commented on tests/components/vacuum/test_xiaomi.py:
```diff
@@ -0,0 +1,161 @@
+    assert state.attributes.get(ATTR_CLEANED_AREA) == 123.43
+    assert state.attributes.get(ATTR_FAN_SPEED) == 'Quiet'
+    assert(state.attributes.get(ATTR_FAN_SPEED_LIST)
+           # == list(sorted(FAN_SPEEDS.keys(), key=lambda k: FAN_SPEEDS[k])))
```
Is this old code or a comment on how the list is sorted?

azogue replied:
both ;-)
When making the test, I realized the list has to be sorted, so I changed it in the Xiaomi code and commented here, to make it fixed in the test. 
Remove?

MartinHjelmare replied:
Ok for me to keep, but perhaps write "The fan speed list is sorted like this:" to clarify?

azogue replied:
ok

azogue replied:
I think I'll remove it

MartinHjelmare commented on tests/components/vacuum/test_xiaomi.py:
```diff
@@ -0,0 +1,161 @@
+    # Call services
+    yield from hass.services.async_call(
+        DOMAIN, SERVICE_TURN_ON, blocking=True)
+    assert(str(mock_mirobo.mock_calls[-2])
```
Is not having a space between `assert` and the first parenthesis when breaking a line following some style guide like PEP, or why not have a space?

azogue replied:
Now I'm seeing it, I think it was also because alignment and legibility ... another old habit. Don't like it?

MartinHjelmare replied:
Well `assert` is not a function in Python but a statement:
https://docs.python.org/3/reference/simple_stmts.html#the-assert-statement

So I think it's more clear to have a space after it.

MartinHjelmare reviewed (commented):
Xiaomi tests look good. Very nice to include tests for that as well. I had a few questions.

azogue committed:
fix xiaomi test

MartinHjelmare commented on homeassistant/components/vacuum/demo.py:
```diff
@@ -0,0 +1,203 @@
+        if self.supported_features & SUPPORT_SEND_COMMAND == 0:
+            return
+
+        self._status = "Executing %s(%s)" % (command, params)
```
```py
self._status = "Executing {}({})".format(command, params)
```
I noticed also when looking at the coverage report that this method wasn't hit by a test. I think're always mocking the `send_command` service, so the method will never get called. You could add a test where you don't mock the service, and test that `self._status` and `self._state` has the correct state after the service call.

azogue replied:
you're right, I forgot. I did it for the `set_fan_speed` but no for `send_command`. I'll do it

MartinHjelmare replied:
Great! Just fix the string formatting on this line and I think we can wrap this.

azogue replied:
you mean `'Executing {}({})'.format(command, params)`??

MartinHjelmare replied:
Yes.

azogue replied:
done

MartinHjelmare commented:
The xiaomi tests are not included in the coverage report, cause all xiaomi platforms are excluded with a wildcard. When that is fixed and my comment above, I think this is done. Awesome job! :tada:

azogue committed:
fix coveragerc

azogue committed:
test send command

azogue commented:
As @pvizeli advised, that has been an adventure!

 🎉 🎉 🎉

MartinHjelmare commented on .coveragerc:
```diff
@@ -197,7 +197,9 @@ omit =
 
     homeassistant/components/xiaomi.py
-    homeassistant/components/*/xiaomi.py
+    homeassistant/components/light/xiaomi.py
```
Add binary_sensor and cover platforms too for exclusion.

azogue replied:
sorry, fixed

pschmitt commented:
You did a truly great job @azogue. Hats off.

azogue committed:
fix coveragerc

azogue committed:
fix string formatting

bramkragten commented:
🎉 Wow! Very nice job @azogue! Thanks!

MartinHjelmare reviewed (approved):
@pvizeli any comments? Ok to merge?

pvizeli committed:
The coverage is to low. It need 93% or more

pvizeli commented:
Realy amazing work. That was what I had want for vacuum component in my first comments :+1:  we have a realy great community. Thanks for that

azogue commented:
I'll look to the xiaomi vacuum coverage in detail to increase it when I'll have the time, didn't see the 87%.

CharlesBlonde commented:
I started working on a vacuum platform in order to add support for Dyson 360 eye robot vacuum (I didn't like to consider a robot vacuum as a switch). It was only a few hours work so it was far far from a production ready platform and thanks to you work I will be able to only focus on implementing it. Thanks !

balloob commented:
This has introduced a flaky test. Seen this one fail a couple of times now: https://travis-ci.org/home-assistant/home-assistant/jobs/261247791#L812

azogue commented:
Saw it, @balloob, I'll fix it later today in other PR
